tutorials/kotlin-springboot-container-engine.md

```diff
@@ -0,0 +1,383 @@
+you can [download](https://github.com/jetbrains/gcp-samples) the sample application 
+
+2. Download the generated project and save it to a local folder
+
+3. Open the resulting project in your favourite IDE or editor and create a new source file named `MessageController.kt` with the following contents
+
+```kotlin
+package com.jetbrains.demo
+
+import org.springframework.web.bind.annotation.*
+
+data class Message(val text: String, val priority: String)
+
+@RestController
+class MessageController {
+    @RequestMapping("/message")
+    fun message(): Message {
+        return Message("Hello from Google Cloud", "High")
+    }
+}
+```
+
+The package should match that of your group and artifact name. 
+
+4. Make sure you have the right dependencies in your Gradle file to import `RestController`
+
+```groovy
+	compile("org.springframework.boot:spring-boot-starter-web")
+```
+
+5. Run the application from the command line using Gradle
+
+    gradle bootRun
+
+
+**Note:** The `gradle bootRun` is a quick way to build and run the application. Later on when creating the Docker image, you'll 
+need to first build the app using the Gradle `build` task and then run it.
+
+6. Open the browser and make sure your get a valid JSON response when accessing
+
+    http://localhost:8080/message
+    
+The result should be
+
+```json
+{
+"text": "Hello from Google Cloud",
+"priority": "High"
+}
+```    
+    
+## Dockerizing your application
+
+The next step is to produce a Docker image that builds and runs your
+application in a Docker container. You will define this image using a
+Dockerfile.
+
+### Create a Dockerfile
+
+Various considerations go into designing a good Docker image. The Dockerfile
+used by this tutorial builds a release and runs it with Alpine Linux.
+If you are experienced with Docker, you can customize your image.
+
+1.  Create a file called `Dockerfile` in your project directory. Copy the
+    following content into it. Alternately, you can
+    [download](https://github.com/JetBrains/gcp-samples/blob/master/Dockerfile)
+    the Dockerfile to study and customize.
+
+        FROM openjdk:8-jdk-alpine
+        VOLUME /tmp
+        RUN mkdir /work
+        COPY . /work 
+        WORKDIR /work
+        RUN /work/gradlew build
+        RUN mv /work/build/libs/*.jar /work/app.jar
+        ENTRYPOINT ["java","-Djava.security.egd=file:/dev/./urandom","-jar","/work/app.jar"]
+
+
+2.  Create a file called `.dockerignore` in your project directory. Copy the
+    following content into it. Alternately, you can
+    [download](https://github.com/JetBrains/gcp-samples/blob/master/.dockerignore)
+    the Dockerfile file to study and customize.
+
+        .gradle
+        build
+        out
+    
+
+The Dockerfile performs a series of steps. It creates a `work` folder where sources will be copied to and built from.
```

Comment:
The Dockerfile performs a series of steps:
1. It creates a `work` folder where sources are copied to and built from.
2. It copies the sources from the local file system (which should be your project folder) to the container.
3. It sets the working directory to the `work` folder created and then runs the Gradle wrapper command (`gradlew`) to build the project. This wrapper automatically downloads Gradle and then builds the project. 
4. Once the project is built, the Dockerfile copies the output of the `jar` to the `work` folder, where it is ready for 
    execution when you invoke `docker run`.